File: betterrandom/src/main/java/io/github/pr0methean/betterrandom/util/LooperThread.java
package io.github.pr0methean.betterrandom.util;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Wraps a thread that loops a given task until interrupted, with the iterations being
 * transactional.
 */
public abstract class LooperThread implements Serializable {

  /**
   * Singleton-ization of {@link Executors#defaultThreadFactory()}.
   */
  protected static final ThreadFactory DEFAULT_THREAD_FACTORY = Executors.defaultThreadFactory();
  private static final long serialVersionUID = -4790652062170305318L;

  /**
   * The thread holds this lock whenever it is running {@link #iterate()}.
   */
  protected final Lock lock = new ReentrantLock(true);
  protected final Lock threadLock = new ReentrantLock();
  @SuppressWarnings("InstanceVariableMayNotBeInitializedByReadObject")
  protected transient volatile Thread thread;
  protected final ThreadFactory factory;
  private volatile boolean running; // determines whether to start when deserialized
  private volatile boolean everStarted; // tracked for getState()

  /**
   * Constructs a LooperThread with all properties as defaults. The thread is not started in the
   * constructor, because subclass fields won't have been initialized.
   */
  protected LooperThread() {
    this(DEFAULT_THREAD_FACTORY);
  }

  /**
   * Constructs a LooperThread with a thread name. The thread is not started in the
   * constructor, because subclass fields won't have been initialized.
   *
   * @param name the name of the thread to create
   */
  protected LooperThread(final String name) {
    this(r -> {
      Thread thread = DEFAULT_THREAD_FACTORY.newThread(r);
      thread.setName(name);
      return thread;
    });
  }

  /**
   * Constructs a LooperThread with a given thread factory. The thread is not started in the
   * constructor, because subclass fields won't have been initialized.
   *
   * @param factory the thread factory that will create this instance's thread
   */
  protected LooperThread(ThreadFactory factory) {
    this.factory = factory;
  }

  public boolean isRunning() {
    threadLock.lock();
    try {
      return thread != null && thread.isAlive();
    } finally {
      threadLock.unlock();
    }
  }

  private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
    in.defaultReadObject();
    if (running) {
      start();
    }
  }

  /**
   * The task that will be iterated until it returns false. Cannot be abstract for serialization
   * reasons, but must be overridden in subclasses if they are instantiated without a target {@link
   * Runnable}.
   *
   * @return true if this thread should iterate again.
   * @throws InterruptedException if interrupted in mid-execution.
   */
  @SuppressWarnings("BooleanMethodIsAlwaysInverted") protected abstract boolean iterate()
      throws InterruptedException;

  protected void start() {
    threadLock.lock();
    try {
      if (thread == null || !thread.isAlive()) {
        thread = factory.newThread(this::run);
        thread.start();
        everStarted = true;
        running = true;
      }
    } finally {
      threadLock.unlock();
    }
  }

  public void interrupt() {
    threadLock.lock();
    try {
      running = false;
      if (thread != null) {
        thread.interrupt();
        thread = null;
      }
    } finally {
      threadLock.unlock();
    }
  }

  @Override public String toString() {
    return thread.toString();
  }

  protected Thread.State getState() {
    threadLock.lock();
    try {
      if (thread == null) {
        return everStarted ? Thread.State.TERMINATED : Thread.State.NEW;
      }
      return thread.getState();
    } finally {
      threadLock.unlock();
    }
  }

  /**
   * Runs {@link #iterate()} until either it returns false or this thread is interrupted.
   */
  private void run() {
    try {
      while (true) {
        lock.lockInterruptibly();
        try {
          if (!iterate()) {
            interrupt();
            return;
          }
        } finally {
          lock.unlock();
        }
      }
    } catch (final InterruptedException ignored) {
      interrupt();
    }
  }

  private void writeObject(ObjectOutputStream out) throws IOException {
    lock.lock();
    try {
      out.defaultWriteObject();
    } finally {
      lock.unlock();
    }
  }
}

File: betterrandom/src/main/java/io/github/pr0methean/betterrandom/util/LooperThread.kt
package io.github.pr0methean.betterrandom.util

import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.locks.Condition
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock

/**
 * Thread that loops a given task until interrupted (or until JVM shutdown, if it [ ][.isDaemon]), with the iterations being transactional.
 */
open class LooperThread : Thread {

    protected val finishedIterations = AtomicLong(0)
    /**
     * The thread holds this lock whenever it is being serialized or cloned or is running [ ][.iterate] called by [.run].
     */
    protected val lock: Lock = ReentrantLock(true)
    protected val endOfIteration = lock.newCondition()
    /**
     * The [Runnable] that was passed into this thread's constructor, if any.
     */
    protected var target: Runnable? = null

    /**
     * Constructs a LooperThread with the given name and target. `target` should only be null if
     * called from a subclass that overrides [.iterate].
     * @param target If not null, the target this thread will run in [.iterate].
     * @param name the thread name
     */
    constructor(target: Runnable?, name: String) : super(target, name) {
        this.target = target
    }

    /**
     * Constructs a LooperThread that belongs to the given [ThreadGroup] and has the given
     * target. `target` should only be null if called from a subclass that overrides [ ][.iterate].
     * @param group The ThreadGroup this thread will belong to.
     * @param target If not null, the target this thread will run in [.iterate].
     */
    constructor(group: ThreadGroup, target: Runnable?) : super(group, target) {
        this.target = target
    }

    /**
     *
     * Constructs a LooperThread with the given name and target, belonging to the given [ ] and having the given preferred stack size. `target` should only be null if
     * called from a subclass that overrides [.iterate].
     *
     * See [Thread.Thread] for caveats about
     * specifying the stack size.
     * @param group The ThreadGroup this thread will belong to.
     * @param target If not null, the target this thread will run in [.iterate].
     * @param name the thread name
     * @param stackSize the desired stack size for the new thread, or zero to indicate that this
     * parameter is to be ignored.
     */
    constructor(group: ThreadGroup, target: Runnable?, name: String,
                stackSize: Long) : super(group, target, name, stackSize) {
        this.target = target
    }

    /**
     * Constructs a LooperThread with the given name and belonging to the given [ThreadGroup].
     * Protected because it does not set a target, and thus should only be used in subclasses that
     * override [.iterate].
     * @param group The ThreadGroup this thread will belong to.
     * @param name the thread name
     */
    protected constructor(group: ThreadGroup, name: String) : super(group, name) {}

    /**
     * Constructs a LooperThread with the given target. `target` should only be null if called
     * from a subclass that overrides [.iterate].
     * @param target If not null, the target this thread will run in [.iterate].
     */
    constructor(target: Runnable?) : super(target) {
        this.target = target
    }

    /**
     * Constructs a LooperThread with the given name and target, belonging to the given [ ]. `target` should only be null if called from a subclass that overrides
     * [.iterate].
     * @param group The ThreadGroup this thread will belong to.
     * @param target If not null, the target this thread will run in [.iterate].
     * @param name the thread name
     */
    constructor(group: ThreadGroup, target: Runnable?, name: String) : super(group, target, name) {
        this.target = target
    }

    /**
     * Constructs a LooperThread with the given name. Protected because it does not set a target, and
     * thus should only be used in subclasses that override [.iterate].
     * @param name the thread name
     */
    protected constructor(name: String) : super(name) {}

    /**
     * Constructs a LooperThread with all properties as defaults. Protected because it does not set a
     * target, and thus should only be used in subclasses that override [.iterate].
     */
    constructor() {}

    /**
     * The task that will be iterated until it returns false. Cannot be abstract for serialization
     * reasons, but must be overridden in subclasses if they are instantiated without a target [ ].
     * @return true if this thread should iterate again.
     * @throws InterruptedException if interrupted in mid-execution.
     * @throws UnsupportedOperationException if this method has not been overridden and [     ][.target] was not set to non-null during construction.
     */
    @Throws(InterruptedException::class)
    protected open fun iterate(): Boolean {
        if (target == null) {
            throw UnsupportedOperationException("This method should be overridden, or else this " + "thread should have been created with a Serializable target!")
        } else {
            target!!.run()
            return true
        }
    }

    /**
     * Runs [.iterate] until either it returns false or this thread is interrupted.
     */
    override fun run() {
        while (true) {
            try {
                lock.lockInterruptibly()
                try {
                    val shouldContinue = iterate()
                    finishedIterations.getAndIncrement()
                    if (!shouldContinue) {
                        break
                    }
                } finally {
                    endOfIteration.signalAll()
                    lock.unlock()
                }
            } catch (ignored: InterruptedException) {
                interrupt()
                break
            }

        }
    }

    /**
     * Wait for the next iteration to finish, with a timeout. May wait longer in the event of a
     * spurious wakeup.
     * @param time the maximum time to wait
     * @param unit the time unit of the `time` argument
     * @return `false`  the waiting time detectably elapsed before an iteration finished, else
     * `true`
     * @throws InterruptedException if thrown by [Condition.await]
     */
    @Throws(InterruptedException::class)
    fun awaitIteration(time: Long, unit: TimeUnit): Boolean {
        val previousFinishedIterations = finishedIterations.get()
        lock.lock()
        try {
            while (!isInterrupted && state != Thread.State.TERMINATED && finishedIterations.get() == previousFinishedIterations) {
                endOfIteration.await(time, unit)
            }
            return finishedIterations.get() != previousFinishedIterations
        } finally {
            lock.unlock()
        }
    }
}
